swupd/cmd/create-pack/main.go

```diff
@@ -0,0 +1,237 @@
+	os.Exit(2)
+}
+
+func main() {
+	log.SetFlags(0)
+
+	cpuProfile := flag.String("cpuprofile", "", "write CPU profile to a file")
+	outputDir := flag.String("o", "", "output directory, creates temporary if not specified")
+	useChroot := flag.Bool("chroot", false, "use chroot to speed up pack generation")
+	allBundles := flag.Bool("all", false, "create packs for all bundles new in TO version")
+	force := flag.Bool("f", false, "rewrite packs that already exist")
+	flag.Usage = usage
+
+	flag.Parse()
+	if (*allBundles && flag.NArg() != 3) || (!*allBundles && flag.NArg() != 4) {
+		usage()
+	}
+
+	if *cpuProfile != "" {
+		f, err := os.Create(*cpuProfile)
+		if err != nil {
+			log.Fatalf("couldn't create file for CPU profile: %s", err)
+		}
+		pprof.StartCPUProfile(f)
+		defer pprof.StopCPUProfile()
+	}
+
+	stateDir := flag.Arg(0)
+
+	fromVersion := flag.Arg(1)
+	toVersion := flag.Arg(2)
+
+	fromVersionUint := parseUint32(fromVersion)
+	toVersionUint := parseUint32(toVersion)
+
+	chrootDir := ""
+	if *useChroot {
+		chrootDir = filepath.Join(stateDir, "image", toVersion, "full")
+		if _, err := os.Stat(chrootDir); err != nil {
+			log.Fatalf("couldn't access the full chroot: %s", err)
+		}
+	}
+
+	if *outputDir == "" {
+		tempDir, err := ioutil.TempDir(".", "packs-")
+		if err != nil {
+			log.Fatalf("couldn't create output directory: %s", err)
+		}
+		*outputDir = tempDir
+	}
+	log.Printf("Output directory: %s", *outputDir)
+
+	// If we are handling a single bundle, its name is taken directly from the command line.
+	if !*allBundles {
+		name := flag.Arg(3)
+		if name == "full" || name == "MoM" || name == "" {
+			log.Fatalf("invalid bundle name %q", name)
+		}
+		bundle := &bundleToPack{
+			name: name,
+			from: fromVersionUint,
+			to:   toVersionUint,
+		}
+		pack(stateDir, chrootDir, *outputDir, bundle, *force)
+		return
+	}
+
+	//
+	// Collect the corresponding bundle versions. Note that the code below will create packs on
+	// different directories, based on the version that each bundle is in the Manifest.MoM for
+	// the specified toVersion.
+	//
+	toDir := filepath.Join(stateDir, "www", toVersion)
+	toMoM, err := readManifestFile(filepath.Join(toDir, "Manifest.MoM"))
+	if err != nil {
+		log.Fatalf("couldn't read MoM of TO_VERSION (%s): %s", toVersion, err)
+	}
+
+	bundles := make(map[string]*bundleToPack, len(toMoM.Files))
+	for _, f := range toMoM.Files {
+		bundles[f.Name] = &bundleToPack{f.Name, 0, f.Version}
+	}
+
+	// If this is not a zero pack, we might be able to skip some bundles.
+	if fromVersionUint > 0 {
+		fromDir := filepath.Join(stateDir, "www", fromVersion)
+		fromMoM, err := readManifestFile(filepath.Join(fromDir, "Manifest.MoM"))
+		if err != nil {
+			log.Fatalf("couldn't read MoM of FROM_VERSION (%s): %s", fromVersion, err)
+		}
+
+		for _, f := range fromMoM.Files {
+			e, ok := bundles[f.Name]
+			if !ok {
+				// Bundle doesn't exist in new version, no pack needed.
+				continue
+			}
+			if e.to == f.Version {
```

Comment:
Added an early distro version check.

We still need this per bundle check here, since different Clear Linux versions might have a bundle in the same version, so we want to skip it for packing purposes.

Reply:
Oh, right. f.Version is for the manifest represented in the MoM, which can definitely have a version==toVersion.

Reply:
actually... the file `f` is from `fromMoM`, which is the `Manifest.MoM` in `fromVersion`, which should be older than `toVersion` (and `e.to==toVersion`), so this case should never be hit because `f.Version` can not be any newer than `fromVersion` (the version of `fromMoM` from whence it came). What am I missing here?

Reply:
disregard again, misread the code.